core/domain/exp_jobs_one_off_test.py

```diff
@@ -939,6 +1058,295 @@ def test_no_action_is_performed_for_deleted_exploration(self):
+            u'[u\'NumericExpressionInput\', [u\'exp_id0 State3: 3.5+4/2\']]']
+
+        self.assertEqual(actual_output, expected_output)
+
+        exploration = exp_fetchers.get_exploration_by_id('exp_id0')
+
+        self.assertEqual(
+            exploration.get_interaction_id_by_state_name('State1'),
+            'AlgebraicExpressionInput')
+        self.assertEqual(
+            exploration.get_interaction_id_by_state_name('State2'),
+            'MathEquationInput')
+        self.assertEqual(
+            exploration.get_interaction_id_by_state_name('State3'),
+            'NumericExpressionInput')
+
+    def test_explorations_with_invalid_inputs_are_not_upgraded(self):
+        """Checks that invalid explorations are not upgraded and an error is
+        yielded.
+        """
+        exploration = exp_domain.Exploration.create_default_exploration(
+            self.VALID_EXP_ID, title='title', category='category')
+
+        exploration.add_states(['State1'])
+
+        state1 = exploration.states['State1']
+
+        state1.update_interaction_id('MathExpressionInput')
+
+        answer_group_list1 = [{
+            'rule_specs': [{
+                'rule_type': 'IsMathematicallyEquivalentTo',
+                'inputs': {'x': 'x<y>z'}
+            }, {
+                'rule_type': 'IsMathematicallyEquivalentTo',
+                'inputs': {'x': 'x+y=c'}
+            }],
+            'outcome': {
+                'dest': 'Introduction',
+                'feedback': {
+                    'content_id': 'feedback',
+                    'html': '<p>Outcome for state1</p>'
+                },
+                'param_changes': [],
+                'labelled_as_correct': False,
+                'refresher_exploration_id': None,
+                'missing_prerequisite_skill_id': None
+            },
+            'training_data': [],
+            'tagged_skill_misconception_id': None
+        }]
+
+        state1.update_interaction_answer_groups(answer_group_list1)
+        exp_services.save_new_exploration(self.albert_id, exploration)
+
+        job_id = (
+            exp_jobs_one_off.MathExpressionUpgradeOneOffJob.create_new())
+        exp_jobs_one_off.MathExpressionUpgradeOneOffJob.enqueue(job_id)
+        self.process_and_flush_pending_tasks()
+
+        actual_output = (
+            exp_jobs_one_off.MathExpressionUpgradeOneOffJob.get_output(
+                job_id))
+        expected_output = [
+            u'[u\'Invalid\', [u\'exp_id0 State1: x<y>z\']]',
+            u'[u\'ERROR\', [u\'There are some invalid inputs that need to be '
+            u'resolved before running the upgrade job.\']]',
+            u'[u\'MathEquationInput\', [u\'exp_id0 State1: x+y=c\']]']
+
+        self.assertEqual(actual_output, expected_output)
+
+        exploration = exp_fetchers.get_exploration_by_id('exp_id0')
+
+        # The exploration contains an invalid input and must therefore, not be
```

Comment:
Validate also that no changes at all have been made to the exploration.

Reply:
I've now checked if the version remains the same or not. That should suffice, right?

Reply:
Mm, sort of. It's just about acceptable, but it would be better to have a more thorough check of the content of the exploration itself since the version number is one step removed.

Reply:
Ok, I've updated the tests to check for the entire exploration content, instead of just the version.